Improvements in the handling of feature compatibility for plugins

Konamiman opened the pull request:
### Changes proposed in this Pull Request:

Changes in the admin UI that handles warnings about incompatibilities between WooCommerce features and plugins, to make the mechanism more flexible and less noisy:

1. Now compatibility won't be checked in the plugins page for features that aren't enabled. For example, if HPOS is disabled (the posts table is selected as the storage for orders), no warnings about plugins incompatible with HPOS will be displayed.
	* However, in the HPOS settings page (_WooCommerce - Settings - Advanced - Features - Order data storage_), nothing changes: a warning will be displayed, and HPOS won't be available for enabling, if at least one incompatible plugin is present and active.

2. The "Incompatible with feature" views of the plugins page now works as follows:
    * When the page is displaying all the incompatible plugins (query string is `plugin_status=incompatible_with_feature`), plugins that are incompatible with any **enabled** feature will be listed.
    * But when the page is displaying plugins that are incompatible with a given feature (query string is `plugin_status=incompatible_with_feature&feature_id=<feature id>`),  then all the plugins that are incompatible with the feature are shown - even if the feature is currently disabled.

HPOS has the flag set to `true`, all the other features don't have it so plugins will be considered compatible with them by default.

3. Now plugins that don't declare any compatibility (positive or negative) with a given feature will be considered _compatible_ with the feature. A new optional key is added to the feature definition array: `plugins_are_incompatible_by_default`; when it's `true`, plugins that don't declare any compatibility with a given feature will be considered _incompatible_ with the feature.

4. A new filter is added, `woocommerce_plugins_are_incompatible_with_feature_by_default`. It takes a feature id and the default (in)compatible by default flag for it, and returns the actual flag value to use.

Additionally, there are changes in the HPOS related CLI commands:

4. The `wp wc hpos enable` command has a new optional argument, `--ignore-plugin-compatibility`, that allows to enable HPOS even if incompatible plugins are active.

5. A new command is added: `wp wc hpos compatibility`. It will list all the WooCommerce-aware plugins (optionally also the inactive ones) in three lists: compatible, incompatible and uncertain. This information is independent from the new `plugins_are_incompatible_by_default` flag.

Closes #47920.

### How to test the changes in this Pull Request:

#### Initial setup

We'll start with a site that has no orders (or where all the orders are in sync between HPOS tables and the post tables) and has no WooCommerce-aware plugins other than WooCommerce itself. There we'll create four plugins, the files can go directly in `wp-content/plugins`:

* `wc-hpos-compatible.php`:

```php
<?php
/**
 * Plugin Name: WooCommerce test - COMPATIBLE with HPOS
 * WC tested up to: 9.0.0
 */

add_action( 'before_woocommerce_init', function() {
    \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, true );
} );
```

* `wc-hpos-incompatible.php`:

```php
<?php
/**
 * Plugin Name: WooCommerce test - INCOMPATIBLE with HPOS
 * WC tested up to: 9.0.0
 */

add_action( 'before_woocommerce_init', function() {
    \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, false);
} );
```

* `wc-hpos-uncertain.php`:

```php
<?php
/**
 * Plugin Name: WooCommerce test - UNCERTAIN about HPOS
 * WC tested up to: 9.0.0
 */
```

* `wc-unaware.php`:

```php
<?php
/**
 * Plugin Name: WooCommerce test - UNAWARE of WooCommerce
 */
```

After creating these, open the following pages in your testing site:

* All plugins page (`/wp-admin/plugins.php`)
* All incompatible plugins (`/wp-admin/plugins.php?plugin_status=incompatible_with_feature`)
* Plugins incompatible with HPOS (`/wp-admin/plugins.php?plugin_status=incompatible_with_feature&feature_id=custom_order_tables`)
* Features settings (`/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`)

Initially all the plugins are disabled, the incompatible plugins list is empty, and you can enable and disable HPOS freely. Make sure that HPOS is enabled and let's begin our testing from this point.

#### The new "incompatible by default" flag

1. Activate all the test plugins. Verify that this causes:
    * The plugins page to display the "WooCommerce has detected that some of your active plugins are incompatible with currently enabled WooCommerce features" warning.
    * The all incompatible plugins page to list two plugin, the "incompatible" one and the "uncertain" one.
    * Same for the plugins incompatible with HPOS page.
    * The HPOS settings section has a warning that says "2 incompatible plugins detected" (and the HPOS setting is grayed out, even if still enabled).

#### The new filter

2. Add the following code snippet to the site:

```php
add_filter(
    'woocommerce_plugins_are_incompatible_with_feature_by_default',
    fn($incompatible_by_default, $feature_id) => $incompatible_by_default && $feature_id !== 'custom_order_tables',
    10, 2);
```

3. Verify that now:
    * Both of the incompatible plugins pages (for all features and for HPOS) now list just one plugin, the "incompatible" one.
    * The warning in the HPOS settings section now says "1 incompatible plugin detected".

4. Remove the snippet.

#### Enabling/disabling features

4. Disable HPOS (select the posts table in "Order data storage" in the features settings page).
5. Verify that the all incompatible plugins page is now empty.
6. Verify that the HPOS incompatible plugins page is still listing the "incompatible" and the "uncertain" plugins.

#### Activating/deactivating plugins

7. Enable HPOS again (spoiler: you can use `wp wc hpos enable --ignore-plugin-compatibility`)
8. Try activating and deactivating the "incompatible" and the "uncertain" plugins, and verify that the incompatible plugins lists update consistently on reload: only the active plugins appear in the lists.
9. The "X incompatible plugins detected" warning should also update accordingly, listing the exact same plugins that are in the incompatible plugins lists (and not displaying any warning if the lists are empty).

#### Changes in CLI

10. Deactivate the "incompatible" plugin, keep the "uncertain" plugin active.
11. Add again the snippet that you used to test the filter.
12. In the command line run `wp wc hpos enable`. It should succeed.
13. Run `wp wc hpos disable` and remove the snippet again.
14. Try again `wp wc hpos enable`. This time it will fail.
15. Run `wp wc hpos enable --ignore-plugin-compatibility`. This time it should succeed.
16. Run `wp wc hpos compatibility-info` and verify that the information shown is accurate. Try with the `--include-inactive` and `--display-filenames` arguments. Note that when using `--include-inactive`, inactive plugins will always appear in the "uncertain" list.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Konamiman committed:
Improvements in the handling of feature compatibility for plugins:

- In the plugins list only features that are currently enabled
  will be considered for the list of features that are incompatible
  with a given plugin.

- A new setting is added to treat uncertain plugins (those that don't
  declare explicit compatibility with a given feature) as incompatible
  (will be disabled by default, which is the opposite of the
  current handling).

Konamiman committed:
Improvements in the CLI runner for HPOS:

- Add a --ignore-plugin-compatibility switch to the "wc hpos enable"
  command, to allow enabling even if there are incompatible plugins.

- Add a new "wc hpos compatibility", which lists the plugins
  that are compatible/incompatible/uncertain with HPOS.

Konamiman committed:
Add changelog file

Konamiman committed:
Merge branch 'trunk' into improve-feature-compatibility-handling

Konamiman committed:
Fix linting issues

Konamiman committed:
Change teh logic of the "incompatible by default" setting

Now it's "compatible by default" instead.

Konamiman committed:
Fix unit test

Konamiman committed:
Rename the 'wc hpos compatibility' CLI command to 'compatibility-info'

Konamiman committed:
Small fixes in the CLI runner

Konamiman committed:
Fix yet more lint issues

barryhughes commented:
Sorry for the delay in getting to this 🙌🏼 

> Now compatibility won't be checked in the plugins page for features that aren't enabled. For example, if HPOS is disabled (the posts table is selected as the storage for orders), no warnings about plugins incompatible with HPOS will be displayed.

Focusing on this bit initially. Scenario:

* HPOS is disabled.
* An HPOS-incompatible plugin is active.

In this situation, there is still a warning in the feature setting screen (expected):

<div align="center"><img width="500" src="https://github.com/woocommerce/woocommerce/assets/3594411/300b4a1f-826e-4a0a-8d8f-eb578cdde776" alt="Feature settings, showing HPOS is disabled" /></div>

However, if I follow the link to "view and manage" the incompatible plugins, then I'm taken to:

`/wp-admin/plugins.php?plugin_status=incompatible_with_feature&feature_id=custom_order_tables`

With this branch, I find that the list table is empty. Whereas, with trunk, it will be populated with a list of incompatible plugins. This seems a little confusing or unexpected?

barryhughes commented:
> A new setting is added, in the WooCommerce - Settings - Advanced - Features - General section, that allows to choose if uncertain plugins (those that are aware of WooCommerce, but don't declare compatibility nor incompatibility with a given feature) are considered compatible with that feature or not.

Talking this out a little before we commit to it:

- With HPOS, we initially decided that plugins *not* declaring compatibility should be regarded as incompatible. This was a risk mitigation strategy. Now that HPOS is stable and in wide use, it makes less sense ... but it *did* make sense at one point.
- This feels (to me) like something that should be defined feature-by-feature, by the feature author, according to the nature and stability of those features at a given point in time, and not something that should be determined by the merchant/site operator for *all* features.

Could it make more sense to specify how incompatibility is assessed during feature declaration? wdyt?

barryhughes reviewed (commented):
(Left some initial questions.)

Konamiman commented:
> With this branch, I find that the list table is empty. Whereas, with trunk, it will be populated with a list of incompatible plugins. This seems a little confusing or unexpected?

I think it makes sense:

* If the feature is disabled, then no plugin is affected by any incompatibility, and thus showing a "this plugin is incompatible with that feature" warning seems like unnecessary noise.
* But if you go to the features page you'll see that the radio button for enabling HPOS is disabled _and_ there's a list of incompatible plugins: this is the page telling you why HPOS can't be enabled. With this information you can decide to disable the offending plugins, or to enable the feature anyway using CLI (with the new `--ignore-plugin-compatibility` arg).

> This feels (to me) like something that should be defined feature-by-feature

Wouldn't this be even more confusing? If I was a site admin it would be surprising for me that some features consider uncertain plugins as compatible and some don't, whereas with an explicit general setting I always know what's going on _and_ I can choose the level of "noise" I want in the plugins page.

barryhughes commented:
> > With this branch, I find that the list table is empty. Whereas, with trunk, it will be populated with a list of incompatible plugins. This seems a little confusing or unexpected?
>
> I think it makes sense: ...

Broadly speaking, so do I, but the specific thing I'm zeroing in on here is that the list of incompatible plugins in the feature screen are accompanied by a link to *view and manage* those same items (see my earlier screenshot).

Why provide this link, if the resulting table is empty?

Konamiman commented:
> Why provide this link, if the resulting table is empty?

Ah ok, you are right, I had missed this particular nuance. Maybe when HPOS is disabled the link should simply open the all plugins page instead? (and the link text could be just "View plugins")

barryhughes commented:
> > This feels (to me) like something that should be defined feature-by-feature
>
> Wouldn't this be even more confusing?

In this change, we're giving up our ability to say that, for some features (like early stage COT/HPOS), the absence of a compatibility declaration means the plugin should be regarded as risky. 

I think retaining that ability is potentially useful for other high risk features we might introduce in the future. Would it be confusing for merchants? Not if we change how we communicate about it (we could simply say, for this category of plugin, that compatibility is *unknown* and therefore we recommend against enabling it with Feature X, rather than that it is outright incompatible). That second part (finessing how we communicate about these cases) doesn't necessarily have to be something we do now, however, because we don't currently have any such high-risk features available.

Beyond that, I don't know that it is fair to ask merchants to make this decision, and to do it in such an ungranular way (every feature is different, and they can have significantly different risk profiles).

Getting feature management right is important, and lots of teams use it, so we could also open this aspect of the discussion up for further input.

barryhughes commented:
> Maybe when HPOS is disabled the link should simply open the all plugins page instead? (and the link text could be just "View plugins")

I think that would be better. The chief thing is to let the merchant see the full plugin details so they can, perhaps, contact the authors and ask about their plans. On this level, I think linking to and populating a filtered view of the relevant plugins is better, but if the LOE is too high for that I think what you are suggesting is fine.

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -1213,6 +1226,104 @@ public function backfill( array $args = array(), array $assoc_args = array() ) {
+		$this->log(
+			sprintf(
+				// translators: $1$d = plugins count, %2$s = colon (if list follows) or empty.
+				__( '%%C%1$d%%n compatible plugins found%2$s', 'woocommerce' ),
```
Here and for similar lines, since there could be zero, one or many plugins found, should we either swap `plugins` for `plugin(s)` or else use [_n()](https://developer.wordpress.org/reference/functions/_n/) (I believe this is the most internationally friendly approach)?

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -1213,6 +1226,104 @@ public function backfill( array $args = array(), array $assoc_args = array() ) {
+	 * @param array $args       Positional arguments passed to the command.
+	 * @param array $assoc_args Associative arguments (options) passed to the command.
+	 */
+	public function compatibility_info( array $args = array(), array $assoc_args = array() ) {
```
(Doesn't require action, this is more a high-level thought) I wonder if it would make most sense to support a family of commands for feature compatibility, rather than potentially registering commands for individual features like HPOS?

That is, rather than having an HPOS specific command:

```sh
wp wc hpos compatibility-info
```

Have a compatibility command that can work with any feature:

```sh
wp wc compatibility <feature-slug>
```

Konamiman replied:
This is certainly a good idea but I think it's better to implement this in a separate PR.

Konamiman commented:
> Could it make more sense to specify how incompatibility is assessed during feature declaration?

Going back to your initial suggestion: you mean a new optional `is_compatible_by_default` boolean flag for each feature, defaulting to `false`, and setting it to `true` for the cart checkout blocks feature (and maybe for any of the other non-legacy features we have)? Sure, that's doable. Maybe we could complement that with a filter (pass the feature id, receive the value of the flag) in case someone _really_ wants to reduce (or increase!) the "noise" in the plugins page.

> Have a compatibility command that can work with any feature

That's a great idea, and we could even take it further and implement a command to enable or disable any feature, as an extension of the `hpos enable/disable` command.

barryhughes commented:
> you mean a new optional is_compatible_by_default boolean flag for each feature, defaulting to false, and setting it to true for the cart checkout blocks feature (and maybe for any of the other non-legacy features we have)?

Yes, that would basically cover it. 

Except (and this is really just an opinion, I don't mind if we go the other way) I think that new flag should default to `true` because, for *most* features, the risk profile probably isn't high enough to warrant treating the absence of a compatibility declaration as meaning it is effectively incompatible.

In other words, features like HPOS—where an initial phase exists in which we do not assume compatibility by default—are the exception.

Though again, whether it is true or false by default matters less: what I think is valuable is having the ability to define this on a feature-by-feature basis.

Konamiman committed:
Fix the "View and manage" link for incompatible plugins.

If the feature is disabled the link will now point to the "All plugins"
page, since the "incompatible with this feature" page would be empty.

Konamiman committed:
Properly pluralize the "plugin/plugins" messages in CLI

Konamiman committed:
Strategy for deciding when "undecided" means "incompatible" changed

Now, instead of a setting affecting all the plugins/features,
a per-feature flag named "plugins_are_incompatible_by_default"
is used. Defaulting to false, when it's true for a given feature
it means that plugins not declaring compatibility with a the
feature are to be considered incompatible with the feature.
The flag is set for the HPOS feature only.

Also a new 'woocommerce_plugins_are_incompatible_with_feature_by_default'
filter is added, which allows to alter the flag for a given feature.

Konamiman commented:
> I think linking to and populating a filtered view of the relevant plugins is better

In the end I modified how the incompatible plugins page works: the "incompatible with single feature" page shows all the incompatible plugins, even for disabled features (as this is always relevant information); the "incompatible with all features" page only shows plugins that are incompatible with enabled features. Sorry if this is starting to sound as a tongue twister.

Konamiman committed:
Fix linting issues

barryhughes commented:
I'm not really sure if this is quite in scope for this PR ... but, going to note it here anyway. Scenario:

- HPOS is disabled.
- Compatibility mode is enabled.
- All orders are synchronized.

<div align="center"><img src="https://github.com/woocommerce/woocommerce/assets/3594411/36dd6adf-e558-435b-a80b-37ec992b5295" alt="HPOS settings, with HPOS disabled and compatibility mode enabled" width="600"/></div>

If another plugin is active, and that plugin is either incompatible with HPOS **or** does not declare compatibility with HPOS, we still prevent the merchant/site operator from enabling HPOS.

I don't know if this is wrong, as such, but especially in the second case (where that other plugin simply has not declared compatibility), it seems excessive. Especially since we allow enabling HPOS via WP CLI

barryhughes commented:
> **The new filter**
>
> Add the following code snippet to the site:
> ...
> Both of the incompatible plugins pages (for all features and for HPOS) now list just one plugin, the "uncertain" one.

They now just list `WooCommerce test - INCOMPATIBLE with HPOS`, which I think is the true and expected result (that is, with this snippet in place, we no longer regard uncertain plugins as being incompatible).

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -7,6 +7,8 @@
 use Automattic\WooCommerce\Internal\DataStores\Orders\LegacyDataHandler;
 use Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore;
 use Automattic\WooCommerce\Internal\Features\FeaturesController;
+use Automattic\WooCommerce\Monorepo\Composer\Plugin;
```
```suggestion
```

Seems like this import is not used?

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -1213,6 +1226,104 @@ public function backfill( array $args = array(), array $assoc_args = array() ) {
+	 * @param array $args       Positional arguments passed to the command.
+	 * @param array $assoc_args Associative arguments (options) passed to the command.
+	 */
+	public function compatibility_info( array $args = array(), array $assoc_args = array() ) {
```
```suggestion
	public function compatibility_info( array $args = array(), array $assoc_args = array() ): void {
```

(Optional) since it's a shiny brand new method, we could include a return type.

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -1213,6 +1226,104 @@ public function backfill( array $args = array(), array $assoc_args = array() ) {
+	 * @param bool  $display_filenames True to simply return the sorted list of plugin file names.
+	 * @return array A sorted array of plugin names or file names.
+	 */
+	private function get_printable_plugin_names( array $plugins, bool $display_filenames ) {
```
```suggestion
	private function get_printable_plugin_names( array $plugins, bool $display_filenames ): array {
```

(Similar to earlier comment, also optional.)

barryhughes commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -322,6 +325,35 @@ function ( $feature ) {
+		 *
+		 * @param bool $incompatible_by_default Default value, true if plugins are to be considered incompatible by default with the feature.
+		 * @param string $feature_id The feature to check.
+		 * @return bool New value.
```
```suggestion
```

Generally, we don't use `@return` when [documenting filters](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#4-hooks-actions-and-filters).

barryhughes commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -322,6 +325,35 @@ function ( $feature ) {
+		 *
+		 * @since 9.2.0
+		 */
+		return apply_filters( 'woocommerce_plugins_are_incompatible_with_feature_by_default', $incompatible_by_default, $feature_id );
```
```suggestion
		return (bool) apply_filters( 'woocommerce_plugins_are_incompatible_with_feature_by_default', $incompatible_by_default, $feature_id );
```

Could be more defensive?

Konamiman commented:
> If another plugin is active, and that plugin is either incompatible with HPOS or does not declare compatibility with HPOS, we still prevent the merchant/site operator from enabling HPOS.

Yes, that's exactly how it's designed and how it's supposed to work: if at least one plugin is incompatible (or is uncertain and the feature is configured to treat those as incompatible, as it's the case of HPOS) then we don't allow to enable HPOS via UI. The admin has two options then: disable the plugin and enable HPOS (and then re-enable the plugin), or use CLI with the new `--ignore-plugin-compatibility` flag. In both cases the admin needs to be intentional about it, so activating HPOS accidentally in incompatibility scenarios isn't possible.

> They now just list WooCommerce test - INCOMPATIBLE with HPOS, which I think is the true and expected result

Indeed, the testing instructions were wrong, sorry.

Konamiman committed:
Small fixes

barryhughes commented:
> Yes, that's exactly how it's designed and how it's supposed to work ...

For sure, I understand that. I'm challenging if that still makes sense, particularly for plugins where incompatibility is assumed, rather than explicitly declared (this isn't something that needs to hold up this PR, however).

barryhughes reviewed (approved):
Seems ready to go 🚀 

Re-checked most of the flows after the final round of small adjustments, nothing left to add.

barryhughes commented:
@Konamiman I approved but cannot merge due to failing CI checks (they seem unrelated to this change). I triggered a re-run so, between us, let's circle back and merge once everything is green.
